<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Guillermo Reales" />

<meta name="date" content="2023-10-13" />

<title>Computing a PGS using RÃ¡pidoPGS-single and GWAS catalog</title>

<script src="data:application/javascript;base64,Ly8gUGFuZG9jIDIuOSBhZGRzIGF0dHJpYnV0ZXMgb24gYm90aCBoZWFkZXIgYW5kIGRpdi4gV2UgcmVtb3ZlIHRoZSBmb3JtZXIgKHRvCi8vIGJlIGNvbXBhdGlibGUgd2l0aCB0aGUgYmVoYXZpb3Igb2YgUGFuZG9jIDwgMi44KS4KZG9jdW1lbnQuYWRkRXZlbnRMaXN0ZW5lcignRE9NQ29udGVudExvYWRlZCcsIGZ1bmN0aW9uKGUpIHsKICB2YXIgaHMgPSBkb2N1bWVudC5xdWVyeVNlbGVjdG9yQWxsKCJkaXYuc2VjdGlvbltjbGFzcyo9J2xldmVsJ10gPiA6Zmlyc3QtY2hpbGQiKTsKICB2YXIgaSwgaCwgYTsKICBmb3IgKGkgPSAwOyBpIDwgaHMubGVuZ3RoOyBpKyspIHsKICAgIGggPSBoc1tpXTsKICAgIGlmICghL15oWzEtNl0kL2kudGVzdChoLnRhZ05hbWUpKSBjb250aW51ZTsgIC8vIGl0IHNob3VsZCBiZSBhIGhlYWRlciBoMS1oNgogICAgYSA9IGguYXR0cmlidXRlczsKICAgIHdoaWxlIChhLmxlbmd0aCA+IDApIGgucmVtb3ZlQXR0cmlidXRlKGFbMF0ubmFtZSk7CiAgfQp9KTsK"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<link rel="stylesheet" href="data:text/css,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" type="text/css" />




</head>

<body>




<h1 class="title toc-ignore">Computing a PGS using RÃ¡pidoPGS-single and GWAS catalog</h1>
<h4 class="author">Guillermo Reales</h4>
<h4 class="date">2023-10-13</h4>



<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>RÃ¡pidoPGS is a tool to quickly compute polygenic scores (PGS) from GWAS summary statistics datasets from either case-control (eg. asthma) or quantitative traits (eg. height and BMI).</p>
<p>Input GWAS summary statistics datasets should have a minimum set of columns, with these names, but in any order:</p>
<ul>
<li><strong>CHR</strong>: Chromosome</li>
<li><strong>BP</strong>: Base position (in GRCh37/hg19 or GRCh38/hg38). If using hg38, use build = âhg38â in parameters.</li>
<li><strong>SNPID</strong>: rsids, or SNP identifiers. If not available, they can be anything (eg. CHR_BP).</li>
<li><strong>REF</strong>: Reference, or non-effect allele</li>
<li><strong>ALT</strong>: Alternative, or effect allele, the one  refers to</li>
<li><strong>BETA</strong>: <span class="math inline">\(\beta\)</span> (or log(OR)), or effect sizes.</li>
<li><strong>SE</strong>: Standard error of <span class="math inline">\(\beta\)</span></li>
<li><strong>P</strong>: P-values for the association test</li>
</ul>
<p>Also, if doing a PGS on a quantitative trait GWAS dataset, your file must include this:</p>
<ul>
<li><strong>ALT_FREQ</strong>: Minor/ALT allele frequency in the tested population, or in a close population from a reference panel.</li>
</ul>
</div>
<div id="installation-of-rÃ¡pidopgs" class="section level1">
<h1>Installation of RÃ¡pidoPGS</h1>
<p>Current RÃ¡pidoPGS version (v2.1.0) is available on CRAN, so we can install it using the following code</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;RapidoPGS&quot;</span>)</span></code></pre></div>
<p>Alternatively, you can download the development version from Github (Note: If you donât have <code>remotes</code> installed, please install it first:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;remotes&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">&quot;remotes&quot;</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;GRealesM/RapidoPGS&quot;</span>)</span></code></pre></div>
</div>
<div id="setup" class="section level1">
<h1>Setup</h1>
<p>Once installed, letâs load it. This will automatically load all required dependencies.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RapidoPGS)</span></code></pre></div>
</div>
<div id="downloading-data-from-gwas-catalog" class="section level1">
<h1>Downloading data from GWAS catalog</h1>
<p><a href="https://www.ebi.ac.uk/gwas/">GWAS catalog</a> is an outstanding GWAS summary statistics data source, as it not only puts together tons of public datasets, but it also uses a semi-automatic pipeline to apply quality control and liftover (a.k.a. harmonise) those datasets, thus helping overcome the problem of having GWAS sumstats in so many different formats and builds.</p>
<p>In this vignette, weâll use GWAS catalog to obtain an example dataset. For this vignette weâll use a Breast cancer (BRCA) dataset from <a href="https://www.nature.com/articles/nature24284">Michailidou et al., 2017</a>, which is one that we used in our manuscript. This GWAS was performed on 119078 controls and 137045 cases of Breast cancer.</p>
<p>Itâs a big file, and may take a while to download, so here we will show the command to download, but actually cheat and load a subset of data already loaded.</p>
<p>Note that in this particular case we chose a study that contained a single dataset. In other cases there may be more than one. In that case <code>gwascat.download</code> will prompt you with the list of available options, prompting their accession numbers, and asking you to choose a file by its number in the list, if running interactively.</p>
<p>We use itâs PubMed ID (29059683). When running <code>gwascat.download</code> without any other arguments, it will try to get the harmonised files associated with the ID. Harmonised files have been processed by GWAS catalog and are formatted and aligned to the lastest genome build. See <a href="https://www.ebi.ac.uk/gwas/docs/methods/summary-statistics">here</a> for more information.</p>
<p>Once we download the file, weâll need to prepare it for RÃ¡pidoPGS. That will involve renaming columns to something RÃ¡pidoPGS can understand, and this is easy to do with data.table. Also, RÃ¡pidoPGS use only autosomes, so remove X or Y chromosomes at this step.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>ds <span class="ot">&lt;-</span> <span class="fu">gwascat.download</span>(<span class="dv">29059683</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Select the harmonised hg38 file </span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="co"># This is equivalent to:</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="co"># ds &lt;- fread(&quot;ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST004001-GCST005000/GCST004988/harmonised/29059683-GCST004988-EFO_0000305.h.tsv.gz&quot;)</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Then apply some reformatting</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="fu">setnames</span>(ds, <span class="at">old =</span> <span class="fu">c</span>(<span class="st">&quot;hm_rsid&quot;</span>,<span class="st">&quot;hm_chrom&quot;</span>,<span class="st">&quot;hm_pos&quot;</span>, <span class="st">&quot;hm_other_allele&quot;</span>, <span class="st">&quot;hm_effect_allele&quot;</span>, <span class="st">&quot;hm_effect_allele_frequency&quot;</span>, <span class="st">&quot;hm_beta&quot;</span>, <span class="st">&quot;standard_error&quot;</span>, <span class="st">&quot;p_value&quot;</span>), <span class="at">new =</span> <span class="fu">c</span>(<span class="st">&quot;SNPID&quot;</span>,<span class="st">&quot;CHR&quot;</span>, <span class="st">&quot;BP&quot;</span>, <span class="st">&quot;REF&quot;</span>,<span class="st">&quot;ALT&quot;</span>,<span class="st">&quot;ALT_FREQ&quot;</span>, <span class="st">&quot;BETA&quot;</span>, <span class="st">&quot;SE&quot;</span>, <span class="st">&quot;P&quot;</span>))</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>ds <span class="ot">&lt;-</span> ds[,.(SNPID, CHR, BP, REF, ALT, ALT_FREQ, BETA, SE, P)]</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>ds <span class="ot">&lt;-</span> ds[CHR <span class="sc">!=</span><span class="st">&quot;X&quot;</span>]</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>ds<span class="sc">$</span>CHR <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(ds<span class="sc">$</span>CHR)</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>ds <span class="ot">&lt;-</span> ds[<span class="fu">order</span>(CHR, BP)]</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>ds <span class="ot">&lt;-</span> <span class="fu">na.omit</span>(ds, <span class="at">cols =</span> <span class="fu">c</span>(<span class="st">&quot;BETA&quot;</span>, <span class="st">&quot;ALT_FREQ&quot;</span>))</span></code></pre></div>
<p>For illustrative purposes, I took a random subset from this file including 100,000 SNPs from this file, which weâll use in this tutorial. Bear in mind that the final PGS wonât be a <em>valid</em> model since we randomly removed most SNPs. It will serve for our teaching purposes, though! ;) We can load this file straight away.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>ds <span class="ot">&lt;-</span> michailidou38</span></code></pre></div>
</div>
<div id="applying-quality-control-to-the-dataset" class="section level1">
<h1>Applying quality control to the dataset</h1>
<p>The first thing to do is to check what our file looks like. RÃ¡pidoPGS can handle NAs in crucial columns, so donât worry too much about that (unless you have all NA for crucial columns, of course!).</p>
<p>A note of caution when dealing with GWAS catalog files, though: since they use a semi-automated pipeline, it is possible that even some columns are present, they are <em>empty</em>, so be careful with that!</p>
<p>Also, RÃ¡pidoPGS requires allele frequencies, so itâs possible that you need to compute it from an external reference panel (eg. 1000 Genomes Phase III). We donât cover that step in this vignette, but we might write instructions explaining how to do it in the future.</p>
<p>Lastly, we applied a number of QC steps in our paper, which we wonât apply here, but encourage you to try when using real data. The details of this procedure are described in the paper. You can also take a look at the code <a href="https://github.com/GRealesM/RapidoPGS_paper/blob/master/Preparing_datasets_qcfilt_clean_20210317.R">here</a>.</p>
<p>Letâs now look at the file.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(ds)</span></code></pre></div>
<pre><code>##     SNPID                CHR               BP                REF           
##  Length:100000      Min.   : 1.000   Min.   :    19484   Length:100000     
##  Class :character   1st Qu.: 4.000   1st Qu.: 32314220   Class :character  
##  Mode  :character   Median : 7.000   Median : 69249390   Mode  :character  
##                     Mean   : 8.612   Mean   : 78711220                     
##                     3rd Qu.:13.000   3rd Qu.:114464115                     
##                     Max.   :22.000   Max.   :248922216                     
##      ALT               ALT_FREQ           BETA                  SE         
##  Length:100000      Min.   :0.0051   Min.   :-2.1291000   Min.   :0.00620  
##  Class :character   1st Qu.:0.0205   1st Qu.:-0.0100000   1st Qu.:0.00740  
##  Mode  :character   Median :0.1037   Median :-0.0003000   Median :0.01140  
##                     Mean   :0.2245   Mean   :-0.0009533   Mean   :0.01949  
##                     3rd Qu.:0.3571   3rd Qu.: 0.0094000   3rd Qu.:0.02540  
##                     Max.   :0.9949   Max.   : 0.6442000   Max.   :0.83960  
##        P                N         
##  Min.   :0.0000   Min.   :256123  
##  1st Qu.:0.1922   1st Qu.:256123  
##  Median :0.4470   Median :256123  
##  Mean   :0.4597   Mean   :256123  
##  3rd Qu.:0.7174   3rd Qu.:256123  
##  Max.   :0.9991   Max.   :256123</code></pre>
<p>In this case, we donât have any missing data, which is fantastic.</p>
</div>
<div id="computing-pgs-using-rÃ¡pidopgs-single" class="section level1">
<h1>Computing PGS using RÃ¡pidoPGS-single</h1>
<p>Letâs now compute our PGS! The build of this example file is <strong>hg38</strong>, so we must tell RÃ¡pidoPGS about that (default is hg19).</p>
<p>In this new version, we donât need sample size for case-control datasets. Note that if this was a quantitative trait dataset, we should inform total number of individuals (N parameter). Also, if our dataset had columns reporting the number of individuals for each SNP, we can replace N by a string specifying the column (eg. N = âsample_sizeâ). By doing so, RÃ¡pidoPGS-single will extract this information directly from the columns.</p>
<p>Letâs get our hands dirty! Letâs compute first a <em>full</em> RÃ¡pidoPGS-single model.</p>
<p><strong>Advanced use note</strong>: You may want to filter by and align your dataset to an external reference. You can do that with RÃ¡pidoPGS, too, by specifying the path of your reference file at <code>reference</code> argument. This reference file should have five columns (CHR, BP, REF, ALT, and SNPID) and be in the <strong>same build</strong> as our summary statistics dataset. RÃ¡pidoPGS currently supports <strong>hg19</strong> and <strong>hg38</strong> builds.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>full_PGS <span class="ot">&lt;-</span> <span class="fu">rapidopgs_single</span>(ds, <span class="at">trait =</span> <span class="st">&quot;cc&quot;</span>, <span class="at">build =</span> <span class="st">&quot;hg38&quot;</span>)</span></code></pre></div>
<pre><code>## Assigning LD blocks...</code></pre>
<pre><code>## Done!</code></pre>
<pre><code>## Computing a RapidoPGS-single model for a case-control dataset...</code></pre>
<p>Well, that was <em>rÃ¡pido</em>! With increasingly big datasets, it will take a bit longer, of course.</p>
<p>Letâs take a look.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(full_PGS)</span></code></pre></div>
<pre><code>##         SNPID CHR        BP REF ALT ALT_FREQ    BETA     SE      P      N
## 1: rs76197520   2  49673981   C   T   0.0603  0.0149 0.0131 0.2540 256123
## 2: rs12232959   2 149575863   C   G   0.7818 -0.0035 0.0076 0.6450 256123
## 3:  rs7621633   3   7330932   C   T   0.7460 -0.0045 0.0089 0.6151 256123
## 4: rs11129616   3  34773314   A   G   0.6731 -0.0060 0.0069 0.3895 256123
## 5: rs62282714   3 150895438   T   C   0.1539 -0.0052 0.0113 0.6439 256123
## 6: rs62375579   5 109299301   G   A   0.1909  0.0080 0.0102 0.4347 256123
##    ld.block          ppi        weight
## 1:      135 1.244593e-05  1.854443e-07
## 2:      178 4.221391e-06 -1.477487e-08
## 3:      221 5.050474e-06 -2.272713e-08
## 4:      235 5.029883e-06 -3.017930e-08
## 5:      287 6.268896e-06 -3.259826e-08
## 6:      457 6.922043e-06  5.537634e-08</code></pre>
<p>We see three new columns: <strong>ld.block</strong> corresponds to the <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4731402/">ld-detect</a> LD block number assigned to the SNP (see manuscript for more details of where this comes from), <strong>ppi</strong> is the posterior probability that the SNP is causal, and <strong>weight</strong> is the weighted effect size (BETA * ppi) - and the column weâre interested in.</p>
<div id="applying-different-thresholds" class="section level2">
<h2>Applying different thresholds</h2>
<p>Imagine that we want a small portable PGS. We could use a threshold (eg. keep only SNPs with ppi larger than <span class="math inline">\(1e^{-4}\)</span>, a reasonable threshold) or keep the top SNPs with largest weights (in either way).</p>
<p>We can do that using RÃ¡pidoPGS, letâs see how by using a <span class="math inline">\(1e^{-4}\)</span> threshold.</p>
<p>For accuracy reasons, we recommend recomputing the PGS on just these SNPs after the threshold was applied, so <code>recalc = TRUE</code> by default.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>PGS_1e4 <span class="ot">&lt;-</span> <span class="fu">rapidopgs_single</span>(ds, <span class="at">trait =</span><span class="st">&quot;cc&quot;</span>, <span class="at">build =</span> <span class="st">&quot;hg38&quot;</span>, <span class="at">filt_threshold =</span> <span class="fl">1e-4</span>)</span></code></pre></div>
<pre><code>## Assigning LD blocks...</code></pre>
<pre><code>## Done!</code></pre>
<pre><code>## Computing a RapidoPGS-single model for a case-control dataset...</code></pre>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(PGS_1e4)</span></code></pre></div>
<pre><code>##          SNPID CHR       BP REF ALT ALT_FREQ   BETA     SE       P      N
## 1: rs143863360  15 75910714   G   T   0.0222 0.0611 0.0265 0.02123 256123
##    ld.block          ppi       weight
## 1:     1112 0.0001789831 1.093587e-05</code></pre>
<p>You can omit recalculation by setting that argument to <code>FALSE</code></p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>PGS_1e4_norecalc <span class="ot">&lt;-</span> <span class="fu">rapidopgs_single</span>(ds, <span class="at">trait =</span><span class="st">&quot;cc&quot;</span>, <span class="at">build =</span> <span class="st">&quot;hg38&quot;</span>, <span class="at">filt_threshold =</span> <span class="fl">1e-4</span>, <span class="at">recalc =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<pre><code>## Assigning LD blocks...</code></pre>
<pre><code>## Done!</code></pre>
<pre><code>## Computing a RapidoPGS-single model for a case-control dataset...</code></pre>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(PGS_1e4_norecalc)</span></code></pre></div>
<pre><code>##          SNPID CHR       BP REF ALT ALT_FREQ   BETA     SE       P      N
## 1: rs143863360  15 75910714   G   T   0.0222 0.0611 0.0265 0.02123 256123
##    ld.block          ppi       weight
## 1:     1112 0.0001789831 1.093587e-05</code></pre>
<p>And what if we want the <em>top 10</em> SNPs? Just change the <code>filt_threshold</code> argument. If itâs larger than 1, RÃ¡pidoPGS will understand that you want a top list, rather than a thresholded result.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>PGS_top10 <span class="ot">&lt;-</span> <span class="fu">rapidopgs_single</span>(ds, <span class="at">trait =</span><span class="st">&quot;cc&quot;</span>, <span class="at">build =</span> <span class="st">&quot;hg38&quot;</span>, <span class="at">filt_threshold =</span> <span class="dv">10</span>)</span></code></pre></div>
<pre><code>## Assigning LD blocks...</code></pre>
<pre><code>## Done!</code></pre>
<pre><code>## Computing a RapidoPGS-single model for a case-control dataset...</code></pre>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(PGS_top10)</span></code></pre></div>
<pre><code>##          SNPID CHR       BP REF ALT ALT_FREQ    BETA     SE       P      N
## 1: rs143863360  15 75910714   G   T   0.0222  0.0611 0.0265 0.02123 256123
## 2:  rs55817933  19 36970383   T   C   0.3584 -0.0154 0.0065 0.01691 256123
## 3: rs149557496   9 35906658   G   A   0.0152  0.0297 0.0313 0.34330 256123
## 4:  rs72867697  17 76595643   T   A   0.1731 -0.0165 0.0095 0.08168 256123
## 5:  rs76197520   2 49673981   C   T   0.0603  0.0149 0.0131 0.25400 256123
## 6:  rs61932677  12 98186323   G   A   0.1930  0.0096 0.0093 0.30170 256123
##    ld.block          ppi        weight
## 1:     1112 1.789831e-04  1.093587e-05
## 2:     1257 5.361020e-05 -8.255971e-07
## 3:      757 2.399329e-05  7.126007e-07
## 4:     1198 2.136836e-05 -3.525779e-07
## 5:      135 1.244593e-05  1.854443e-07
## 6:      978 7.904310e-06  7.588138e-08</code></pre>
</div>
</div>
<div id="conclusion" class="section level1">
<h1>Conclusion</h1>
<p>So those are examples of basic <em>RÃ¡pidoPGS-single</em> usage!</p>
<p>Drop us a line if you encounter problems, and weâll be happy to help.</p>
<p>Good luck!</p>
<p>Guillermo</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
