<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Guillermo Reales" />

<meta name="date" content="2022-06-15" />

<title>Computing a PGS using RápidoPGS-single and GWAS catalog</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Computing a PGS using RápidoPGS-single and
GWAS catalog</h1>
<h4 class="author">Guillermo Reales</h4>
<h4 class="date">2022-06-15</h4>



<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>RápidoPGS is a tool to quickly compute polygenic scores (PGS) from
GWAS summary statistics datasets from either case-control (eg. asthma)
or quantitative traits (eg. height and BMI).</p>
<p>Input GWAS summary statistics datasets should have a minimum set of
columns, with these names, but in any order:</p>
<ul>
<li><strong>CHR</strong>: Chromosome</li>
<li><strong>BP</strong>: Base position (in GRCh37/hg19 or GRCh38/hg38).
If using hg38, use build = “hg38” in parameters.</li>
<li><strong>SNPID</strong>: rsids, or SNP identifiers. If not available,
they can be anything (eg. CHR_BP).</li>
<li><strong>REF</strong>: Reference, or non-effect allele</li>
<li><strong>ALT</strong>: Alternative, or effect allele, the one refers
to</li>
<li><strong>BETA</strong>: <span class="math inline">\(\beta\)</span>
(or log(OR)), or effect sizes.</li>
<li><strong>SE</strong>: Standard error of <span class="math inline">\(\beta\)</span></li>
<li><strong>P</strong>: P-values for the association test</li>
</ul>
<p>Also, if doing a PGS on a quantitative trait GWAS dataset, your file
must include this:</p>
<ul>
<li><strong>ALT_FREQ</strong>: Minor/ALT allele frequency in the tested
population, or in a close population from a reference panel.</li>
</ul>
</div>
<div id="installation-of-rápidopgs" class="section level1">
<h1>Installation of RápidoPGS</h1>
<p>Current RápidoPGS version (v2.1.0) is available on CRAN, so we can
install it using the following code</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;RapidoPGS&quot;</span>)</span></code></pre></div>
<p>Alternatively, you can download the development version from Github
(Note: If you don’t have <code>remotes</code> installed, please install
it first:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;remotes&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">&quot;remotes&quot;</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;GRealesM/RapidoPGS&quot;</span>)</span></code></pre></div>
</div>
<div id="setup" class="section level1">
<h1>Setup</h1>
<p>Once installed, let’s load it. This will automatically load all
required dependencies.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RapidoPGS)</span></code></pre></div>
</div>
<div id="downloading-data-from-gwas-catalog" class="section level1">
<h1>Downloading data from GWAS catalog</h1>
<p><a href="https://www.ebi.ac.uk/gwas/">GWAS catalog</a> is an
outstanding GWAS summary statistics data source, as it not only puts
together tons of public datasets, but it also uses a semi-automatic
pipeline to apply quality control and liftover (a.k.a. harmonise) those
datasets, thus helping overcome the problem of having GWAS sumstats in
so many different formats and builds.</p>
<p>In this vignette, we’ll use GWAS catalog to obtain an example
dataset. For this vignette we’ll use a Breast cancer (BRCA) dataset from
<a href="https://www.nature.com/articles/nature24284">Michailidou et
al., 2017</a>, which is one that we used in our manuscript. This GWAS
was performed on 119078 controls and 137045 cases of Breast cancer.</p>
<p>It’s a big file, and may take a while to download, so here we will
show the command to download, but actually cheat and load a subset of
data already loaded.</p>
<p>Note that in this particular case we chose a study that contained a
single dataset. In other cases there may be more than one. In that case
<code>gwascat.download</code> will prompt you with the list of available
options, prompting their accession numbers, and asking you to choose a
file by its number in the list, if running interactively. If you want to
run <code>gwascat.download</code> automatically, and you know the number
already, you can supply it using <code>filenum</code> argument, so R
won’t bug you.</p>
<p>We use it’s PubMed ID (29059683). We could download only harmonised
GWAS catalog columns (thus having a lighter file) but I don’t advise
this because sometimes we may need columns from the original file. We
then set <code>hm_only = FALSE</code> to get all columns.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>ds <span class="ot">&lt;-</span> <span class="fu">gwascat.download</span>(<span class="dv">29059683</span>, <span class="at">hm_only =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>Now for the cheat: for illustrative purposes, I took a random subset
from this file including 100,000 SNPs from this file, which we’ll use in
this tutorial. Bear in mind that the final PGS won’t be a <em>valid</em>
model since we randomly removed most SNPs. It will serve for our
teaching purposes, though! ;)</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>ds <span class="ot">&lt;-</span> michailidou</span></code></pre></div>
</div>
<div id="applying-quality-control-to-the-dataset" class="section level1">
<h1>Applying quality control to the dataset</h1>
<p>The first thing to do is to check what our file looks like. RápidoPGS
can handle NAs in crucial columns, so don’t worry too much about that
(unless you have all NA for crucial columns, of course!).</p>
<p>A note of caution when dealing with GWAS catalog files, though: since
they use a semi-automated pipeline, it is possible that even some
columns are present, they are <em>empty</em>, so be careful with
that!</p>
<p>Also, RápidoPGS requires allele frequencies, so it’s possible that
you need to compute it from an external reference panel (eg. 1000
Genomes Phase III). We don’t cover that step in this vignette, but we
might write instructions explaining how to do it in the future.</p>
<p>Let’s take a look at the file!</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(ds)</span></code></pre></div>
<pre><code>##    hm_rsid            hm_chrom             hm_pos          hm_other_allele   
##  Length:100000      Length:100000      Min.   :    10519   Length:100000     
##  Class :character   Class :character   1st Qu.: 31928836   Class :character  
##  Mode  :character   Mode  :character   Median : 69324516   Mode  :character  
##                                        Mean   : 78427892                     
##                                        3rd Qu.:114421523                     
##                                        Max.   :248906719                     
##  hm_effect_allele      hm_beta           hm_effect_allele_frequency
##  Length:100000      Min.   :-2.9353000   Min.   :0.0051            
##  Class :character   1st Qu.:-0.0099000   1st Qu.:0.0200            
##  Mode  :character   Median :-0.0003000   Median :0.1033            
##                     Mean   :-0.0008408   Mean   :0.2230            
##                     3rd Qu.: 0.0095000   3rd Qu.:0.3537            
##                     Max.   : 0.6549000   Max.   :0.9949            
##  standard_error      p_value      
##  Min.   :0.0062   Min.   :0.0000  
##  1st Qu.:0.0075   1st Qu.:0.1940  
##  Median :0.0115   Median :0.4492  
##  Mean   :0.0197   Mean   :0.4617  
##  3rd Qu.:0.0259   3rd Qu.:0.7215  
##  Max.   :1.0917   Max.   :0.9993</code></pre>
<p>In this case, we don’t have any missing data, which is fantastic.
We’ll now prepare the dataset for RápidoPGS. We need to rename columns
to something RápidoPGS can understand, and this is easy to do with
data.table.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">setnames</span>(ds, <span class="at">old =</span> <span class="fu">c</span>(<span class="st">&quot;hm_rsid&quot;</span>, <span class="st">&quot;hm_chrom&quot;</span>, <span class="st">&quot;hm_pos&quot;</span>, <span class="st">&quot;hm_other_allele&quot;</span>, <span class="st">&quot;hm_effect_allele&quot;</span>, <span class="st">&quot;hm_beta&quot;</span>, <span class="st">&quot;hm_effect_allele_frequency&quot;</span>, <span class="st">&quot;standard_error&quot;</span>, <span class="st">&quot;p_value&quot;</span>), <span class="at">new=</span><span class="fu">c</span>(<span class="st">&quot;SNPID&quot;</span>,<span class="st">&quot;CHR&quot;</span>, <span class="st">&quot;BP&quot;</span>,<span class="st">&quot;REF&quot;</span>, <span class="st">&quot;ALT&quot;</span>, <span class="st">&quot;BETA&quot;</span>, <span class="st">&quot;ALT_FREQ&quot;</span>, <span class="st">&quot;SE&quot;</span>, <span class="st">&quot;P&quot;</span>))</span></code></pre></div>
<p>Now another thing to consider: RápidoPGS use only autosomes, so
remove X or Y chromosomes at this step.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>ds <span class="ot">&lt;-</span> ds[CHR <span class="sc">!=</span> <span class="st">&quot;X&quot;</span>,]</span></code></pre></div>
</div>
<div id="computing-pgs-using-rápidopgs-single" class="section level1">
<h1>Computing PGS using RápidoPGS-single</h1>
<p>Now we have applied QC to our dataset, we can now compute our PGS! We
can easily do that applying the next command, but bear in mind that,
since we used harmonised columns, the build is <strong>hg38</strong>, so
we must tell RápidoPGS about that.</p>
<p>In this new version, we don’t need sample size for case-control
datasets. Note that if this was a quantitative trait dataset, we should
inform total number of individuals (N parameter). Also, if our dataset
had columns reporting the number of individuals for each SNP, we can
replace N by a string specifying the column (eg. N = “sample_size”). By
doing so, RápidoPGS-single will extract this information directly from
the columns.</p>
<p>Let’s get our hands dirty! Let’s compute first a <em>full</em>
RápidoPGS-single model. Since we used harmonised coordinates from GWAS
catalog, our file is in build hg38, and we need to specify it to
RápidoPGS (default is hg19).</p>
<p><strong>Advanced use note</strong>: You may want to filter by and
align your dataset to an external reference. You can do that with
RápidoPGS, too, by specifying the path of your reference file at
<code>reference</code> argument. This reference file should have five
columns (CHR, BP, REF, ALT, and SNPID) and be in the <strong>same
build</strong> as our summary statistics dataset. RápidoPGS currently
supports <strong>hg19</strong> and <strong>hg38</strong> builds.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>full_PGS <span class="ot">&lt;-</span> <span class="fu">rapidopgs_single</span>(ds, <span class="at">trait =</span> <span class="st">&quot;cc&quot;</span>, <span class="at">build =</span> <span class="st">&quot;hg38&quot;</span>)</span></code></pre></div>
<pre><code>## Assigning LD blocks...</code></pre>
<pre><code>## Done!</code></pre>
<pre><code>## Computing a RapidoPGS-single model for a case-control dataset...</code></pre>
<p>Well, that was <em>rápido</em>! With increasingly big datasets, it
will take a bit longer, of course.</p>
<p>Let’s take a look.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(full_PGS)</span></code></pre></div>
<pre><code>##          SNPID CHR        BP REF ALT    BETA ALT_FREQ     SE       P ld.block
## 1: rs117214991   9 109388299   C   T  0.0929   0.0052 0.0559 0.09647      788
## 2:  rs61113083  11  82946128   G   A -0.0336   0.0623 0.0132 0.01086      915
## 3: rs118065272  12  71684434   G   A  0.0021   0.0401 0.0219 0.92530      978
## 4:   rs6067900  20  51693556   T   C -0.0038   0.2610 0.0072 0.59860     1314
## 5:   rs6132000  20  17808883   C   T -0.0119   0.0981 0.0119 0.31710     1302
## 6: rs544195892   3 196463480 TAC   T  0.0221   0.0526 0.0167 0.18470      314
##             ppi        weight
## 1: 9.681417e-05  8.994036e-06
## 2: 1.654291e-04 -5.558417e-06
## 3: 1.091763e-05  2.292701e-08
## 4: 3.651311e-06 -1.387498e-08
## 5: 9.688627e-06 -1.152947e-07
## 6: 1.980177e-05  4.376191e-07</code></pre>
<p>We see three new columns: <strong>ld.block</strong> corresponds to
the <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4731402/">ld-detect</a>
LD block number assigned to the SNP (see manuscript for more details of
where this comes from), <strong>ppi</strong> is the posterior
probability that the SNP is causal, and <strong>weight</strong> is the
weighted effect size (BETA * ppi) - and the column we’re interested
in.</p>
<div id="applying-different-thresholds" class="section level2">
<h2>Applying different thresholds</h2>
<p>Imagine that we want a small portable PGS. We could use a threshold
(eg. keep only SNPs with ppi larger than <span class="math inline">\(1e^{-4}\)</span>, a reasonable threshold) or keep
the top SNPs with largest weights (in either way).</p>
<p>We can do that using RápidoPGS, let’s see how by using a <span class="math inline">\(1e^{-4}\)</span> threshold.</p>
<p>For accuracy reasons, we recommend recomputing the PGS on just these
SNPs after the threshold was applied, so <code>recalc = TRUE</code> by
default.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>PGS_1e4 <span class="ot">&lt;-</span> <span class="fu">rapidopgs_single</span>(ds, <span class="at">trait =</span><span class="st">&quot;cc&quot;</span>, <span class="at">build =</span> <span class="st">&quot;hg38&quot;</span>, <span class="at">filt_threshold =</span> <span class="fl">1e-4</span>)</span></code></pre></div>
<pre><code>## Assigning LD blocks...</code></pre>
<pre><code>## Done!</code></pre>
<pre><code>## Computing a RapidoPGS-single model for a case-control dataset...</code></pre>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(PGS_1e4)</span></code></pre></div>
<pre><code>##          SNPID CHR       BP REF ALT    BETA ALT_FREQ     SE        P ld.block
## 1:  rs61113083  11 82946128   G   A -0.0336   0.0623 0.0132 0.010860      915
## 2:  rs55762932  17 33783485   C   T -0.0839   0.0066 0.0457 0.066720     1198
## 3:  rs17826142  16 48507898   T   C  0.0503   0.0276 0.0195 0.009998     1160
## 4:   rs7277048  21 17703035   C   T -0.0660   0.0124 0.0330 0.045260     1324
## 5: rs141121949   4 42132514   G   A -0.0491   0.0399 0.0167 0.003400      338
## 6: rs139406060   5  4690696   C   T -0.0766   0.0118 0.0404 0.058160      418
##             ppi        weight
## 1: 0.0001656217 -5.564889e-06
## 2: 0.0001100407 -9.232416e-06
## 3: 0.0002614005  1.314845e-05
## 4: 0.0001139903 -7.523357e-06
## 5: 0.0006078236 -2.984414e-05
## 6: 0.0001113266 -8.527615e-06</code></pre>
<p>You can omit recalculation by setting that argument to
<code>FALSE</code></p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>PGS_1e4_norecalc <span class="ot">&lt;-</span> <span class="fu">rapidopgs_single</span>(ds, <span class="at">trait =</span><span class="st">&quot;cc&quot;</span>, <span class="at">build =</span> <span class="st">&quot;hg38&quot;</span>, <span class="at">filt_threshold =</span> <span class="fl">1e-4</span>, <span class="at">recalc =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<pre><code>## Assigning LD blocks...</code></pre>
<pre><code>## Done!</code></pre>
<pre><code>## Computing a RapidoPGS-single model for a case-control dataset...</code></pre>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(PGS_1e4_norecalc)</span></code></pre></div>
<pre><code>##          SNPID CHR       BP REF ALT    BETA ALT_FREQ     SE        P ld.block
## 1:  rs61113083  11 82946128   G   A -0.0336   0.0623 0.0132 0.010860      915
## 2:  rs55762932  17 33783485   C   T -0.0839   0.0066 0.0457 0.066720     1198
## 3:  rs17826142  16 48507898   T   C  0.0503   0.0276 0.0195 0.009998     1160
## 4:   rs7277048  21 17703035   C   T -0.0660   0.0124 0.0330 0.045260     1324
## 5: rs141121949   4 42132514   G   A -0.0491   0.0399 0.0167 0.003400      338
## 6: rs139406060   5  4690696   C   T -0.0766   0.0118 0.0404 0.058160      418
##             ppi        weight
## 1: 0.0001654291 -5.558417e-06
## 2: 0.0001098264 -9.214438e-06
## 3: 0.0002606787  1.311214e-05
## 4: 0.0001138967 -7.517184e-06
## 5: 0.0006072546 -2.981620e-05
## 6: 0.0001112250 -8.519836e-06</code></pre>
<p>And what if we want the <em>top 10</em> SNPs? Just change the
<code>filt_threshold</code> argument. If it’s larger than 1, RápidoPGS
will understand that you want a top list, rather than a thresholded
result.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>PGS_top10 <span class="ot">&lt;-</span> <span class="fu">rapidopgs_single</span>(ds, <span class="at">trait =</span><span class="st">&quot;cc&quot;</span>, <span class="at">build =</span> <span class="st">&quot;hg38&quot;</span>, <span class="at">filt_threshold =</span> <span class="dv">10</span>)</span></code></pre></div>
<pre><code>## Assigning LD blocks...</code></pre>
<pre><code>## Done!</code></pre>
<pre><code>## Computing a RapidoPGS-single model for a case-control dataset...</code></pre>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(PGS_top10)</span></code></pre></div>
<pre><code>##         SNPID CHR        BP REF ALT    BETA ALT_FREQ     SE         P ld.block
## 1: rs79054216  11  69371625   C   T  0.2309   0.0145 0.0265 3.041e-18      909
## 2: rs10788196  10 121658121   T   G -0.1324   0.8495 0.0085 5.935e-55      867
## 3: rs28475635   4 174921828   G   A -0.1030   0.1177 0.0098 5.256e-26      405
## 4:   rs580057   3  27303612   G   A  0.1038   0.5249 0.0062 9.185e-63      236
## 5: rs17756147  14  68569410   G   A -0.0973   0.2310 0.0074 2.634e-39     1082
## 6: rs12997141   2 217071663   C   T -0.0914   0.2349 0.0073 1.690e-35      204
##    ppi  weight
## 1:   1  0.2309
## 2:   1 -0.1324
## 3:   1 -0.1030
## 4:   1  0.1038
## 5:   1 -0.0973
## 6:   1 -0.0914</code></pre>
</div>
</div>
<div id="conclusion" class="section level1">
<h1>Conclusion</h1>
<p>So those are examples of basic <em>RápidoPGS-single</em> usage!</p>
<p>Drop us a line if you encounter problems, and we’ll be happy to
help.</p>
<p>Good luck!</p>
<p>Guillermo</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
